marlowe-runtime/sync/Language/Marlowe/Runtime/Sync/Database/PostgreSQL/GetTransactions.hs

```diff
@@ -0,0 +1,136 @@
+import Language.Marlowe.Protocol.Query.Types
+import Language.Marlowe.Runtime.ChainSync.Api (TxId(..), TxOutRef(..))
+import qualified Language.Marlowe.Runtime.ChainSync.Api as TxOutRef (TxOutRef(..))
+import Language.Marlowe.Runtime.Core.Api
+  ( ContractId(..)
+  , MarloweVersion(..)
+  , MarloweVersionTag(..)
+  , Payout(..)
+  , Transaction(..)
+  , TransactionOutput(..)
+  , TransactionScriptOutput(..)
+  )
+import Language.Marlowe.Runtime.Sync.Database.PostgreSQL.GetContractState (PayoutRow, decodePayout)
+import Language.Marlowe.Runtime.Sync.Database.PostgreSQL.GetTransaction (decodeTransaction)
+import Prelude hiding (init)
+
+getTransactions :: ContractId -> T.Transaction (Maybe SomeTransactions)
+getTransactions (ContractId createTxOutRef@TxOutRef{..}) = runMaybeT do
+  let params = (unTxId txId, fromIntegral txIx)
+
+  _ <- MaybeT $ T.statement params
+    [maybeStatement|
+      SELECT txId :: bytea FROM marlowe.createTxOut WHERE txId = $1 :: bytea AND txIx = $2 :: smallint
+    |]
+
+  txs <- lift $ Map.fromDistinctAscList . V.toList . fmap decodeTransaction <$> T.statement params
+    [vectorStatement|
+      SELECT
+        applyTx.txId :: bytea,
+        (ARRAY_AGG(applyTx.createTxId))[1] :: bytea,
+        (ARRAY_AGG(applyTx.createTxIx))[1] :: smallint,
+        (ARRAY_AGG(applyTx.inputTxId))[1] :: bytea,
+        (ARRAY_AGG(applyTx.inputTxIx))[1] :: smallint,
+        (ARRAY_AGG(applyTx.outputTxIx))[1] :: smallint?,
+        (ARRAY_AGG(applyTx.metadata))[1] :: bytea?,
+        (ARRAY_AGG(applyTx.inputs))[1] :: bytea,
+        (ARRAY_AGG(applyTx.invalidBefore))[1] :: timestamp,
+        (ARRAY_AGG(applyTx.invalidHereafter))[1] :: timestamp,
+        (ARRAY_AGG(applyTx.slotNo))[1] :: bigint,
+        (ARRAY_AGG(applyTx.blockId))[1] :: bytea,
+        (ARRAY_AGG(applyTx.blockNo))[1] :: bigint,
+        (ARRAY_AGG(contractTxOut.rolesCurrency))[1] :: bytea?,
+        (ARRAY_AGG(contractTxOut.state))[1] :: bytea?,
+        (ARRAY_AGG(contractTxOut.contract))[1] :: bytea?,
+        (ARRAY_AGG(txOut.address))[1] :: bytea?,
+        (ARRAY_AGG(txOut.lovelace))[1] :: bigint?,
+        ARRAY_REMOVE(ARRAY_AGG(txOutAsset.policyId), NULL) :: bytea[],
+        ARRAY_REMOVE(ARRAY_AGG(txOutAsset.name), NULL) :: bytea[],
+        ARRAY_REMOVE(ARRAY_AGG(txOutAsset.quantity), NULL) :: bigint[]
+      FROM marlowe.applyTx
+      LEFT JOIN marlowe.contractTxOut
+        ON contractTxOut.txId = applyTx.txId
+        AND contractTxOut.txIx = applyTx.outputTxIx
+      LEFT JOIN marlowe.txOut
+        ON txOut.txId = applyTx.txId
+        AND txOut.txIx = applyTx.outputTxIx
+      LEFT JOIN marlowe.txOutAsset
+        ON txOutAsset.txId = applyTx.txId
+        AND txOutAsset.txIx = applyTx.outputTxIx
+      WHERE applyTx.createTxId = $1 :: bytea
+        AND applyTx.createTxId = $2 :: smallint
+      GROUP BY applyTx.txId
+      ORDER BY (ARRAY_AGG(applyTx.inputTxId))[1], (ARRAY_AGG(applyTx.inputTxIx))[1]
+    |]
+
```

Comment:
I think we'll need performance testing on `mainnet`. I've noticed that some asset-related queries perform poorly there, and this query looks superficially similar to some queries that I've had difficulty with.

Reply:
Yes for sure - FYI I always analyze the queries with `EXPLAIN` to make sure the indexes are being used properly, there are no expensive sequential scans or sorts happening, etc. All the queries here have very good asymptotic complexity. But you're right, we'll want to check on a large network.